File: booster-spring-parent/booster-web-starter/src/main/java/io/github/booster/config/thread/ThreadPoolSetting.java
package io.github.booster.config.thread;

import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.ToString;

/**
 * Thread pool settings
 */
@ToString
@Getter
@EqualsAndHashCode
public class ThreadPoolSetting {

    /**
     * Default pool core size
     */
    public static final int DEFAULT_CORE_SIZE = 10;

    /**
     * Default pool max size
     */
    public static final int DEFAULT_MAX_SIZE = 20;

    /**
     * Default pool queue size
     */
    public static final int DEFAULT_QUEUE_SIZE = 100;

    private int coreSize = DEFAULT_CORE_SIZE;

    private int maxSize = DEFAULT_MAX_SIZE;

    private int queueSize = DEFAULT_QUEUE_SIZE;

    private String prefix;

    /**
     * Default constructor with default values.
     */
    public ThreadPoolSetting() {
    }

    /**
     * Sets pool core size
     * @param coreSize core size need to be greater than 0, or defaults to 10
     */
    public void setCoreSize(int coreSize) {
        this.coreSize = coreSize <= 0 ? DEFAULT_CORE_SIZE : coreSize;
    }

    /**
     * Sets max size of the thread pool
     * @param maxSize max size of thread pool, need to be greater than 0 or defaults to 20
     */
    public void setMaxSize(int maxSize) {
        this.maxSize = maxSize <= 0 ? DEFAULT_MAX_SIZE : maxSize;
    }

    /**
     * Sets queue size of the thread pool
     * @param queueSize queue size, need to be greater than 0 or defaults to 100
     */
    public void setQueueSize(int queueSize) {
        this.queueSize = queueSize <= 0 ? DEFAULT_QUEUE_SIZE : queueSize;
    }

    /**
     * Sets prefix of thread pool
     * @param prefix prefix for thread pool
     */
    public void setPrefix(String prefix) {
        this.prefix = prefix;
    }
}

File: booster-spring-parent/booster-web-starter/src/main/kotlin/io/github/booster/config/thread/ThreadPoolSetting.kt
package io.github.booster.config.thread

import lombok.ToString

const val DEFAULT_CORE_SIZE = 10
const val DEFAULT_MAX_SIZE = 20
const val DEFAULT_QUEUE_SIZE = 100

@ToString
data class ThreadPoolSetting (
    val coreSize: Int?,
    val maxSize: Int?,
    val queueSize: Int?,
    val prefix: String?
) {
    fun validate(name: String): ValidatedThreadPoolSetting {
        var coreSize = if (this.coreSize != null && this.coreSize > 0) {
            this.coreSize
        } else {
            DEFAULT_CORE_SIZE
        }
        var maxSize = if (this.maxSize != null && this.maxSize > 0) {
            this.maxSize
        } else {
            DEFAULT_MAX_SIZE
        }
        var queueSize = if (this.queueSize != null && this.queueSize > 0) {
            this.queueSize
        } else {
            DEFAULT_QUEUE_SIZE
        }
        val prefix = if (this.prefix?.isNotBlank() == true) {
            this.prefix
        } else {
            name
        }

        if (coreSize > maxSize) {
            coreSize = maxSize.apply { maxSize = coreSize }
        }

        return ValidatedThreadPoolSetting(
            coreSize, maxSize, queueSize, prefix
        )
    }
}

data class ValidatedThreadPoolSetting (
    val coreSize: Int,
    val maxSize: Int,
    val queueSize: Int,
    val prefix: String
)
